File: Blind 75/13 - 1D Dynamic Programming/JavaScript/300-longest-increasing-subsequence.js
const { performance } = require('perf_hooks');

class Solution {
  method(nums) {
    let cache = Array.from({ length: nums.length }, () => 1);

    for (let i = nums.length - 1; i >= 0; i--) {
      for (let j = i + 1; j < nums.length; j++) {
        if (nums[i] < nums[j]) cache[i] = Math.max(cache[i], 1 + cache[j]);
      }
    }

    return Math.max(...cache);
  }

  reference(nums) {
    const subsequence = this.logarithmicSort(nums); /* Time O(N * log(N) */

    return subsequence.length;
  }

  logarithmicSort = (nums, subsequence = []) => {
    for (const num of nums) {
      /* Time O(N) */
      const max = subsequence[subsequence.length - 1];

      const canAdd = max < num;
      if (canAdd) {
        subsequence.push(num);
        continue;
      } /* Space O(N) */

      const index = this.binarySearch(num, subsequence); /* Time O(log(N)) */

      subsequence[index] = num;
    }

    return subsequence;
  };

  binarySearch = (num, subsequence) => {
    let [left, right] = [0, subsequence.length - 1];

    while (left < right) {
      /* Time O(log(N)) */
      const mid = (left + right) >> 1;
      const guess = subsequence[mid];

      const isNumTarget = num === guess;
      if (isNumTarget) return mid;

      const isNumGreater = guess < num;
      if (isNumGreater) left = mid + 1;

      const isNumLess = num < guess;
      if (isNumLess) right = mid;
    }

    return left;
  };

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input));
        else this.method(input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input));
        else this.reference(input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  [10, 9, 2, 5, 3, 7, 101, 18],
  [0, 1, 0, 3, 2, 3],
  [7, 7, 7, 7, 7, 7, 7],
];
test.quantify(testCases);

File: Blind 75/13 - 1D Dynamic Programming/300-longest-increasing-subsequence.py
'''
Given an integer array nums, return the length of the longest strictly increasing subsequence.

A subsequence is a sequence that can be derived from an array by deleting some or no elements without changing the order of the remaining elements. For example, [3,6,2,7] is a subsequence of the array [0,3,1,6,2,2,7].
'''
from time import time
from typing import List


class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        sub = []
        for n in nums:
            # Binary search + DP
            if len(sub) == 0 or n > sub[-1]:
                sub.append(n)
            else:
                # Find the index of the smallest number >= n
                idx = 0
                for i in range(len(sub)):
                    if sub[i] >= n:
                        idx = i
                        break
                # Replace that number with n
                sub[idx] = n
        return len(sub)

    def reference(self, nums: List[int]) -> int:
        LIS = [1] * len(nums)

        for i in range(len(nums) - 1, -1, -1):
            for j in range(i + 1, len(nums)):
                if nums[i] < nums[j]:
                    LIS[i] = max(LIS[i], 1 + LIS[j])
        return max(LIS)

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.lengthOfLIS(case))
                else:
                    self.lengthOfLIS(case)
        print(f'Runtime for our solution: {time() - sol_start}\n')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case))
                else:
                    self.reference(case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        [10, 9, 2, 5, 3, 7, 101, 18],
        [0, 1, 0, 3, 2, 3],
        [7, 7, 7, 7, 7, 7, 7],
        # Additional
        [4, 10, 4, 3, 8, 9],
    ]
    test.quantify(test_cases)
